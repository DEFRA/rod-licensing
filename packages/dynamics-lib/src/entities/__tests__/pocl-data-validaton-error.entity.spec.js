import { PoclDataValidationError } from '../pocl-data-validation-error.entity.js'
import { retrieveGlobalOptionSets } from '../..'

describe('pocl staging exception entity', () => {
  let optionSetData
  beforeAll(async () => {
    optionSetData = await retrieveGlobalOptionSets().cached()
  })
  describe('maps from dynamics', () => {
    let exception
    beforeAll(() => {
      exception = PoclDataValidationError.fromResponse(
        {
          '@odata.etag': 'W/"56351087"',
          defra_poclvalidationerrorid: '91f15d18-0aa4-ea11-a812-000d3a64905b',
          defra_firstname: 'Daniel',
          defra_name: 'Ricciardo',
          defra_organisation: 'Fishy Endeavours',
          defra_premises: '14 Howecroft Court',
          defra_street: 'Eastmead Lane',
          defra_town: 'Bristol',
          defra_postcode: 'BS9 1HJ',
          defra_country: 'GB',
          defra_birthdate: '1989-07-01',
          defra_emailaddress: 'daniel-ricc@example.com',
          defra_mobilenumber: '07722 123456',
          defra_preferredmethodofnewsletter: 910400003,
          defra_preferredmethodofconfirmation: 910400000,
          defra_preferredmethodofreminder: 910400002,
          defra_concessions: '[{"type":"Blue Badge","referenceNumber":"123456789"}]',
          defra_startdate: '2021-06-15',
          defra_serialnumber: '14345-48457J',
          defra_permitid: 'test-permit-id',
          defra_transactiondate: '2020-01-01T14:00:00Z',
          defra_amount: 30,
          defra_paymentsource: 'Post Office Sales',
          defra_channelid: '948594',
          defra_methodofpayment: 910400001,
          defra_status: 910400000,
          defra_datasource: 910400000
        },
        optionSetData
      )
    })

    it('returns a PoclValidationError instance', () => {
      expect(exception).toBeInstanceOf(PoclDataValidationError)
    })

    it('has the expected data', () => {
      expect(exception).toMatchSnapshot()
    })
  })

  it('maps to dynamics', async () => {
    const validationError = new PoclDataValidationError()
    validationError.firstName = 'Daniel'
    validationError.lastName = 'Ricciardo'
    validationError.organisation = 'Fishy Endeavours'
    validationError.premises = '14 Howecroft Court'
    validationError.street = 'Eastmead Lane'
    validationError.town = 'Bristol'
    validationError.postcode = 'BS9 1HJ'
    validationError.country = 'GB'
    validationError.birthDate = '1989-07-01'
    validationError.email = 'daniel-ricc@example.com'
    validationError.mobilePhone = '07722 123456'
    validationError.preferredMethodOfConfirmation = optionSetData.defra_preferredcontactmethod.options['910400000']
    validationError.preferredMethodOfNewsletter = optionSetData.defra_preferredcontactmethod.options['910400003']
    validationError.preferredMethodOfReminder = optionSetData.defra_preferredcontactmethod.options['910400002']
    validationError.concessions = '[{"type":"Blue Badge","referenceNumber":123456789}]'
    validationError.startDate = '2021-06-15'
    validationError.serialNumber = '14345-48457J'
    validationError.permitId = 'test-permit-id'
    validationError.transactionDate = '2020-01-01T14:00:00Z'
    validationError.amount = 30
    validationError.paymentSource = 'Post Office Sales'
    validationError.channelId = '948594'
    validationError.methodOfPayment = optionSetData.defra_paymenttype.options['910400001']
    validationError.status = optionSetData.defra_poclvalidationerrorstatus.options['910400000']
    validationError.dataSource = optionSetData.defra_datasource.options['910400000']

    const dynamicsEntity = validationError.toRequestBody()
    expect(dynamicsEntity).toMatchSnapshot()
  })
})
