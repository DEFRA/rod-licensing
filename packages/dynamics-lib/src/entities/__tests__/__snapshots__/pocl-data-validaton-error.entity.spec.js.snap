// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pocl staging exception entity maps from dynamics has the expected data 1`] = `
Object {
  "amount": 30,
  "birthDate": "1989-07-01",
  "channelId": "948594",
  "concessions": "[{\\"type\\":\\"Blue Badge\\",\\"referenceNumber\\":\\"123456789\\"}]",
  "country": "GB",
  "dataSource": Object {
    "description": "Post Office Sales",
    "id": 910400000,
    "label": "Post Office Sales",
  },
  "email": "daniel-ricc@example.com",
  "firstName": "Daniel",
  "id": "91f15d18-0aa4-ea11-a812-000d3a64905b",
  "lastName": "Ricciardo",
  "methodOfPayment": Object {
    "description": "Cash",
    "id": 910400001,
    "label": "Cash",
  },
  "mobilePhone": "07722 123456",
  "organisation": "Fishy Endeavours",
  "paymentSource": "Post Office Sales",
  "permitId": "test-permit-id",
  "postcode": "BS9 1HJ",
  "preferredMethodOfConfirmation": Object {
    "description": "Email",
    "id": 910400000,
    "label": "Email",
  },
  "preferredMethodOfNewsletter": Object {
    "description": "Prefer not to be contacted",
    "id": 910400003,
    "label": "Prefer not to be contacted",
  },
  "preferredMethodOfReminder": Object {
    "description": "Text",
    "id": 910400002,
    "label": "Text",
  },
  "premises": "14 Howecroft Court",
  "serialNumber": "14345-48457J",
  "startDate": "2021-06-15",
  "status": Object {
    "description": "Needs Review",
    "id": 910400000,
    "label": "Needs Review",
  },
  "street": "Eastmead Lane",
  "town": "Bristol",
  "transactionDate": "2020-01-01T14:00:00Z",
}
`;

exports[`pocl staging exception entity maps to dynamics 1`] = `
Object {
  "defra_amount": 30,
  "defra_birthdate": "1989-07-01",
  "defra_channelid": "948594",
  "defra_concessions": "[{\\"type\\":\\"Blue Badge\\",\\"referenceNumber\\":123456789}]",
  "defra_country": "GB",
  "defra_datasource": 910400000,
  "defra_emailaddress": "daniel-ricc@example.com",
  "defra_firstname": "Daniel",
  "defra_methodofpayment": 910400001,
  "defra_mobilenumber": "07722 123456",
  "defra_name": "Ricciardo",
  "defra_organisation": "Fishy Endeavours",
  "defra_paymentsource": "Post Office Sales",
  "defra_permitid": "test-permit-id",
  "defra_postcode": "BS9 1HJ",
  "defra_preferredmethodofconfirmation": 910400000,
  "defra_preferredmethodofnewsletter": 910400003,
  "defra_preferredmethodofreminder": 910400002,
  "defra_premises": "14 Howecroft Court",
  "defra_serialnumber": "14345-48457J",
  "defra_startdate": "2021-06-15",
  "defra_status": 910400000,
  "defra_street": "Eastmead Lane",
  "defra_town": "Bristol",
  "defra_transactiondate": "2020-01-01T14:00:00Z",
}
`;
